cmd/provisioner-localpv/app/helper.go

```diff
@@ -0,0 +1,177 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+This code was taken from https://github.com/rancher/local-path-provisioner
+and modified to work with the configuration options used by OpenEBS
+*/
+
+package app
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	CmdTimeoutCounts = 120
+)
+
+// getPathAndNodeForPV inspects the PV spec to determine the host path used
+//  and the node (via the NodeAffinity) on which host path exists.
+func (p *Provisioner) getPathAndNodeForPV(pv *v1.PersistentVolume) (path, node string, err error) {
+	defer func() {
+		err = errors.Wrapf(err, "failed to delete volume %v", pv.Name)
+	}()
+
+	hostPath := pv.Spec.PersistentVolumeSource.HostPath
+	if hostPath == nil {
+		return "", "", fmt.Errorf("no HostPath set")
+	}
+	path = hostPath.Path
+
+	nodeAffinity := pv.Spec.NodeAffinity
+	if nodeAffinity == nil {
+		return "", "", fmt.Errorf("no NodeAffinity set")
+	}
+	required := nodeAffinity.Required
+	if required == nil {
+		return "", "", fmt.Errorf("no NodeAffinity.Required set")
+	}
+
+	node = ""
+	for _, selectorTerm := range required.NodeSelectorTerms {
+		for _, expression := range selectorTerm.MatchExpressions {
+			if expression.Key == KeyNode && expression.Operator == v1.NodeSelectorOpIn {
+				if len(expression.Values) != 1 {
+					return "", "", fmt.Errorf("multiple values for the node affinity")
+				}
+				node = expression.Values[0]
+				break
+			}
+		}
+		if node != "" {
+			break
+		}
+	}
+	if node == "" {
+		return "", "", fmt.Errorf("cannot find affinited node")
+	}
+	return path, node, nil
+}
+
+// createCleanupPod launches a helper(busybox) pod, to delete the host path.
+//  This porivsioner expects that the host paths are created using
+//  an unique PV path - under a given BasePath. From the absolute path,
+//  it extracs the base path and the PV path. The helper pod is then launched
+//  by mounting the base path - and performing a delete on the unique PV path.
+func (p *Provisioner) createCleanupPod(cmdsForPath []string, name, path, node string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "failed to cleanup volume %v", name)
+	}()
+	if name == "" || path == "" || node == "" {
+		return fmt.Errorf("invalid empty name or path or node")
+	}
+
+	//Validate that non-root directories are not passed for delete
+	// and also perform white/black list validations.
+	config := &CASConfigPVC{}
```

Comment:
CASConfigPVC & path seems to be not related.

It might be better to have following where hostpath is first class citizen
```go
// pkg/hostpath/v1alpha1/path.go

type HostPath string

type Predicate func(HostPath) bool

func IsNonRoot() Predicate {
  return func(HostPath) bool {
    // .. logic
  }
}

func IsWhiteListed() Predicate {
  return func(HostPath) bool {
    // .. logic
  }
}

func IsNotBlackListed() Predicate {
  return func(HostPath) bool {
    // .. logic
  }
}

type Builder struct {
  path HostPath
  checks map[*Predicate]string
}

func NewBuilder() *Builder {
  return &Builder{}
}

func (b *Builder) WithPath(path string) *Builder {
  b.path = HostPath(path)
  return b
}

func (b *Builder) WithChecks(checks ...Predicate) *Builder {
  for _, check := range checks {
    b.WithCheck(p)
  }

  return b
}

func (b *Builder) WithCheck(check Predicate) *Builder {
  return b.WithCheckf(check, "")
}

func (b *Builder) WithCheckf(check Predicate, msg string, args ...interface{}) *Builder {
  b.checks[&check] = fmt.Sprintf(msg, args...)
  return b
}

func (b *Builder) Validate() error {
  if b.path == "" {
    return errors.New("failed to validate: missing host path")
  }

  failedValidations []string
  for check, msg := range b.checks {
    if !check(b.path) {
      failedValidations = append(failedValidations, msg)
    }
  }

  if len(failedValidations) > 0 {
    return errors.Errorf("host path validation failed: %v", failedValidations)
  }
  return nil
}

func (b *Builder) GetParentAndVolumeDir() (string, string, error) {
  err := b.Validate()
  if err != nil {
    return "", "", errors.Wrapf(err, "failed to get parent, volume directories")
  }
  // .. logic
}

```

Reply:
done.